cpl/module_cap_cpl.F90

```diff
@@ -80,28 +79,36 @@ end subroutine diagnose_cplFields
   !-----------------------------------------------------------------------------
 
   ! This subroutine requires ESMFv8 - for coupled FV3
-    subroutine State_RWFields_tiles(state,filename,timeslice,rc)
+    subroutine State_RWFields_tiles(state,filename,rc)
 
       type(ESMF_State), intent(in)          :: state
       character(len=*), intent(in)          :: fileName
-      integer, intent(in)                   :: timeslice
       integer, intent(out)                  :: rc
 
-      ! local
-      type(ESMF_Field)                       :: firstESMFFLD
-      type(ESMF_Field),allocatable           :: flds(:)
-      type(ESMF_GridComp) :: IOComp
-      type(ESMF_Grid) :: gridFv3
-
-      character(len=256) :: msgString
-      integer                                :: i, icount, ifld
+      ! local variables
+      type(ESMF_Array)                       :: array
+      type(ESMF_Grid)                        :: grid
+      type(ESMF_FieldBundle)                 :: fieldbundle
+      type(ESMF_Field), allocatable          :: flds(:)
+      type(ESMF_DistGrid)                    :: distgrid
+      integer                                :: i, icount, ifld, id
       integer                                :: fieldcount, firstfld
+      integer                                :: fieldDimCount, gridDimCount, dimCount, tileCount, ungriddedDimCount
       character(64), allocatable             :: itemNameList(:), fldNameList(:)
       type(ESMF_StateItem_Flag), allocatable :: typeList(:)
+      integer, allocatable                   :: minIndexPTile(:,:), maxIndexPTile(:,:)
+      integer, allocatable                   :: ungriddedLBound(:), ungriddedUBound(:)
+      integer, allocatable                   :: fieldDimLen(:)
+      character(len=32), allocatable         :: gridded_dim_labels(:), ungridded_dim_labels(:)
 
-      character(len=*),parameter :: subname='(module_cap_cpl:State_RWFields_tiles)'
+      character(16), parameter :: convention = 'NetCDF'
+      character(16), parameter :: purpose = 'FV3'
 
-      ! local variables
+      integer, parameter :: max_n_axes = 4
+      integer, parameter :: max_n_dim = 16
```

Comment:
May I ask what the two variables mean? max_n_axes is 4 because it covers (x,y,z,time)? what about max_n_dim? is it all the possible dimensions used in the output file?

Reply:
Yes. The max_n_dim is number possible dimensions labels for a given rank (1-4) where each label has different count.. This is typically needed for "ungridded  dimensions" like 'zaxis', when we can have different element counts for different fields, like 127, 128, (for model level data) and 4 and/or 9 for soil fields. Local array axes_dimcount is used to map axes id (1,2,3 or 4) and count to the id of an dimension that has the same count, which is then used to create ESMF_ATT_GRIDDED_DIM_LABELS  and ESMF_ATT_UNGRIDDED_DIM_LABELS attributes.